### PR TITLE
feat(document): support crawl preset document

### PR DESCRIPTION
**Pull request by xxs-wallace**
one of https://github.com/TabbyML/tabby/pull/2898

### CODE REVIEW

File: ee/tabby-db/src/web_documents.rs [hunk 1]
```diff
@@ -58,6 +58,27 @@ impl DbConn {
...
+
+    pub async fn deactivate_preset_web_document(&self, name: String) -> Result<i64> {
+        let res = query_scalar!("SELECT id FROM web_documents WHERE name = ?;", name)
+            .fetch_one(&self.pool)
+            .await?;
+
+        self.delete_web_document(res).await?;
+        Ok(res)
+    }
+
```
Comment (wsxiaoys) [inline ee/tabby-db/src/web_documents.rs:81]:
```suggestion
```

As discussed: set an preset webdocument to active = `create`. set preset webdocument to inactive = `delete`.

File: ee/tabby-webserver/src/service/web_documents.rs [hunk 2]
```diff
@@ -1,21 +1,42 @@
...
 
 pub fn create(db: DbConn, job_service: Arc<dyn JobService>) -> impl WebDocumentService {
-    WebDocumentServiceImpl { db, job_service }
+    let data: serde_json::Value = serde_json::from_str(PRESET_WEB_DOCUMENTS_DATA).unwrap();
+    let mut preset_web_documents = HashMap::new();
+    for doc in data.as_array().unwrap() {
+        let name = doc.get("name").unwrap();
+        let url = doc.get("crawlerStart").unwrap();
+        preset_web_documents.insert(name.to_string(), url.to_string());
+    }
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/web_documents.rs:28]:
```suggestion
```

In our previous discussion, we've concluded that inactive data shall not being inserted to db.

File: ee/tabby-db/src/web_documents.rs [hunk 3]
```diff
@@ -58,6 +58,16 @@ impl DbConn {
         Ok(res.last_insert_rowid())
     }
 
+    pub async fn deactivate_preset_web_document(&self, name: String) -> Result<()> {
+        let res = query!("DELETE FROM web_documents WHERE name = ?;", name)
+            .execute(&self.pool)
+            .await?;
+        if res.rows_affected() != 1 {
+            return Err(anyhow!("No integration access token to delete"));
```
Comment (wsxiaoys) [inline ee/tabby-db/src/web_documents.rs:66]: need update error message?

File: ee/tabby-schema/src/schema/web_documents.rs [hunk 4]
```diff
@@ -12,17 +12,17 @@ pub struct CustomWebDocument {
...
     pub id: ID,
     pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
     pub job_info: JobInfo,
 }
 
 #[derive(GraphQLObject)]
 #[graphql(context = Context)]
 pub struct PresetWebDocument {
-    pub name: String,
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/web_documents.rs:21]: need to keep `name` field - from user's perspective of view, they treat `id` as opaque string, without knowing its content.

File: ee/tabby-schema/src/schema/mod.rs [hunk 5]
```diff
@@ -596,7 +596,7 @@ impl Query {
         before: Option<String>,
         first: Option<i32>,
         last: Option<i32>,
-        active: bool,
+        include_inactive: bool,
```
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-schema/src/schema/mod.rs:599]: ping

**Review by wsxiaoys — COMMENTED** [review]
Please add unit test for methods of `WebDocumentServiceImpl`.

Otherwise LG

File: ee/tabby-schema/src/schema/web_documents.rs (hunk 4 above)
Comment (xxs-wallace) [reply ee/tabby-schema/src/schema/web_documents.rs:21]: done

File: ee/tabby-db/src/web_documents.rs (hunk 3 above)
Comment (xxs-wallace) [reply ee/tabby-db/src/web_documents.rs]: It is a mistake, I have fixed

File: ee/tabby-schema/src/schema/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xxs-wallace) [reply ee/tabby-schema/src/schema/mod.rs]: fixed

**Pull request merged**